pkg/operator/controller/externaldns/deployment_test.go

```diff
@@ -3015,6 +3015,39 @@ func TestExternalDNSDeploymentChanged(t *testing.T) {
 				}),
 			}),
 		},
+		{
+			description: "if externalDNS has extra volume",
+			expect:      false,
+			originalDeployment: testDeploymentWithVolumes(
+				testSecretVolume("testcreds", "testsecret", "creds", "/run/secrets"),
+				testConfigMapVolume("extravolume", "extra", "files", "/etc"),
+			),
+			mutate: func(dep *appsv1.Deployment) {
+				dep.Spec.Template.Spec.Volumes = []corev1.Volume{testSecretVolume("testcreds", "testsecret", "creds", "/run/secrets")}
```

Comment:
Did you mean to append a new volume here?

Reply:
The testcase is weirdly designed. It make my head turn a little every time I approach it. Which is a sign to rewrite it.

`originalDeployment` is current state, `mutate` function changes the current state into the desired deployment. Then the current and desired deployments are passed to `externalDNSDeploymentChanged` which says if the current state needs to be updated and to which state (`expectedDeployment`).

So here, we have 2 volumes in the current (cluster) state and 1 volume as desired by the operator and we expect no changes to the current state as we now allow unsolicited volumes.

Reply:
👍🏼

Reply:
Actually, you now allow unsolicited containers.  Is it unsolicited volumes are allowed as well as unsoliciated containers?  And neither of them are marked as changes.

Reply:
> we expect no changes to the current state as we now allow unsolicited volumes.
But shouldn't you expect the expectedDeployment to only have one volume?

Reply:
You start with 2 volumes - `testcreds` and `extravolume`.  Then you mutate the volumes to become just one volume `testcreds`.  But your expected deployment still desires the original 2 volumes.  I don't understand the test case - you don't accept the requested change and that is ok?

Reply:
- I start with 2 volumes `testcreds` and `extravolume` - that the current state.
- Then I mutate to have just `testcreds` - the mutation is to get the desired state (the one which is supposed to be enforced by the operator).
- The expected (output) deployment still desires 2 volumes because the current state has all the desired volumes(`testcreds`) and we accept having `extravolume`. BTW this test case should have been implemented at the time of the volumes change but I missed it so added it here even though it doesn't concern any code change I did in the PR.

I believe it's the mutation which misleads you here. The mutation here is to create the desired state from the current. As I already mentioned the test case is a total mislead, we just need to have 3 fields: `current` `desired` and `expected` and remove all the mutation logic which only confuses everything.

Reply:
Added a `TODO` to refactor the testcase.

Reply:
The `tc.mutate` function is not confusing or misleading, it's a common pattern.  You take the current, mutate it, then compare it to the expected state after the mutation.  The part I question is why doesn't the mutation become permanent in the expected outcome?

You mentioned: "The expected (output) deployment still desires 2 volumes because the current state has all the desired volumes(testcreds) and we accept having extravolume.", but if you have applied the mutate function to the current set of volumes, it should have removed that volume.  The expected state should only have the single volume you set in the mutate function.

Reply:
@Miciah could you take a quick look at this test case and let me know if you agree?

Reply:
> You take the current, mutate it, then compare it to the expected state after the mutation

No. The mutation creates the desired (by the operator) state.

> but if you have applied the mutate function to the current set of volumes, it should have removed that volume. The expected state should only have the single volume you set in the mutate function.

Desired deployment != expected deployment. The desired deployment is what the operator desires to see. The expected deployment is what we expect to be in the cluster after the operator made the reconciliation. The enforcement of the desired state is done in the soft manner - only the volumes mentioned in the desired deployment has to be present, the rest is left intact. This test case demonstrates exactly this: we have 2 volumes in the cluster and one of them is exactly as desired, so we `expect` no change (`false`). `expectedDepolyment` field here is set but it's not even used as we only check it if `expect` is `true`.

Reply:
Andrey's explanations in https://github.com/openshift/external-dns-operator/pull/176#discussion_r1002094852 and https://github.com/openshift/external-dns-operator/pull/176#discussion_r1022873335 make sense to me.  `original` represents whatever is in etcd, `tc.mutate(original.DeepCopy())` represents what the operator expects, and `tc.expect` is the result of reconciling those two things, and since the goal is for the operator to ignore volumes and sidecar containers that some external agent may inject, `tc.expect` should include any additional sidecars or volumes.  

> The unsolicited volumes were accepted even before this PR, that was done to cover the cases when, for instance, the kube api token is added by the api server.

Those only get injected into the pods, not into the deployment's pod template spec.  

> The testcase is weirdly designed. It makes my head turn a little every time I approach it.

This whole discussion reminds me of another: https://github.com/openshift/cluster-ingress-operator/pull/611#discussion_r630455754

Reply:
> > The unsolicited volumes were accepted even before this PR, that was done to cover the cases when, for instance, the kube api token is added by the api server.

> 
> Those only get injected into the pods, not into the deployment's pod template spec.

Yes, you are right, I saw the api token volume in the POD, not in the deployment.  Do you think we don't need to be that soft for the volumes?
I personally like the concept of the hard enforcement only of those things which can change the behavior of the operated PODs like `args` `command`, `image` etc. The volumes, in my opinion, can be approached softer - letting the cluster user to add some if it's needed.

> > The testcase is weirdly designed. It makes my head turn a little every time I approach it.

> 
> This whole discussion reminds me of another: https://github.com/openshift/cluster-ingress-operator/pull/611#discussion_r630455754

I agree more with Steve's ordering but that's more religious than factual. Let remove the TODO comment though and add your explanation to the test case, I think it helps understand the idea behind.

Reply:
> Yes, you are right, I saw the api token volume in the POD, not in the deployment. Do you think we don't need to be that soft for the volumes?

I don't think we need to be that permissive about volumes, but it's fine to leave it permissive in this PR (which is focused on allowing sidecars).  

> I personally like the concept of the hard enforcement only of those things which can change the behavior of the operated PODs like `args` `command`, `image` etc. The volumes, in my opinion, can be approached softer - letting the cluster user to add some if it's needed.

The risk of loose enforcement is that someone might start modifying the deployment in unexpected ways, and if some future change in external-dns-operator breaks that customization (for example, adding a conflicting volume with the same name), then that person will complain.  However, we can leave things permissive for the purposes of getting this PR merged.